File: python/src/github_wiki_organiser/sidebar.py
from typing import Any

from .application import Application


class Sidebar(Application):
    """Generates the wiki _Sidebar.md as a nested list of owners/categories
    and their wiki pages."""

    def __init__(self, **options: Any) -> None:
        super().__init__(**options)
        self._wiki_list: list[str] = []

    # private

    def _run(self) -> None:
        self._update_wiki_list()
        with open(self._path_to_sidebar, 'w', encoding='utf-8') as f:
            f.write(''.join(self._wiki_list))

    def _update_wiki_list(self) -> None:
        for namespace, wikis in self._owned_wiki_maps().items():
            self._append_section(namespace, wikis, owned=True)
        for namespace, wikis in self._plain_wiki_maps().items():
            self._append_section(namespace, wikis, owned=False)

    def _append_section(self, namespace: str, wikis: list[str],
                        owned: bool) -> None:
        self._wiki_list.append(
            f'{self._section_heading(namespace, owned=owned)}\n')
        for wiki in wikis:
            self._wiki_list.append(f'  - [[{wiki.replace(".md", "")}]]\n')

    def _section_heading(self, namespace: str, owned: bool) -> str:
        owner_base_url = self._config.owner_base_url
        if owned and owner_base_url:
            url = owner_base_url + namespace.replace('@', '')
            return f'- [{namespace}]({url})'
        return f'- {namespace}'

File: RubyGem/lib/spreen_wiki/sidebar.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'application'

module SpreenWiki
  # Generates the wiki _Sidebar.md as a nested list of owners/categories and
  # their wiki pages.
  class Sidebar < Application
    # @rbs return: void
    def run
      update_wiki_list
      File.write(path_to_sidebar, wiki_list.join)
    end

    private

    # @rbs return: Array[String]
    def wiki_list
      @wiki_list ||= []
    end

    # @rbs return: void
    def update_wiki_list
      owned_wiki_maps.each { |namespace, wikis| append_section(namespace, wikis, owned: true) }
      plain_wiki_maps.each { |namespace, wikis| append_section(namespace, wikis, owned: false) }
    end

    # @rbs namespace: String
    # @rbs wikis: Array[String]
    # @rbs owned: bool
    # @rbs return: void
    def append_section(namespace, wikis, owned:)
      wiki_list << "#{section_heading(namespace, owned:)}\n"
      wikis.each { |wiki| wiki_list << "  - [[#{wiki.gsub('.md', '')}]]\n" }
    end

    # @rbs namespace: String
    # @rbs owned: bool
    # @rbs return: String
    def section_heading(namespace, owned:)
      owner_base_url = config.owner_base_url
      return "- #{namespace}" unless owned && owner_base_url

      "- [#{namespace}](#{owner_base_url + namespace.delete('@')})"
    end
  end
end
